src/System.Net.Http/tests/FunctionalTests/HttpRetryProtocolTests.cs

```diff
@@ -0,0 +1,174 @@
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Http.Functional.Tests
+{
+    public class HttpRetryProtocolTests : HttpClientTestBase
+    {
+        private static readonly string s_simpleContent = "Hello World\r\n";
+        private static readonly string s_simpleResponse =
+            $"HTTP/1.1 200 OK\r\n" +
+            $"Date: {DateTimeOffset.UtcNow:R}\r\n" +
+            $"Content-Length: {s_simpleContent.Length}\r\n" +
+            "\r\n" +
+            s_simpleContent;
+
+        // Retry logic is supported by managed handler, CurlHandler, uap, and netfx.  Only WinHttp does not support. 
+        private bool IsRetrySupported =>
+            UseManagedHandler || !PlatformDetection.IsWindows || PlatformDetection.IsUap || PlatformDetection.IsFullFramework;
+
+        public static Task CreateServerAndClientAsync(Func<Uri, Task> clientFunc, Func<Socket, Task> serverFunc)
+        {
+            IPEndPoint ignored;
+            return LoopbackServer.CreateServerAsync(async (server, uri) =>
+            {
+                Task clientTask = clientFunc(uri);
+                Task serverTask = serverFunc(server);
+
+                await TestHelper.WhenAllCompletedOrAnyFailed(clientTask, serverTask);
+
+            }, out ignored);
+        }
+
+        [Fact]
+        public async Task GetAsync_RetryOnConnectionClosed_Success()
+        {
+            if (!IsRetrySupported)
+            {
+                return;
+            }
+
+            await CreateServerAndClientAsync(async url =>
+            {
+                using (HttpClient client = CreateHttpClient())
+                {
+                    // Send initial request and receive response so connection is established
+                    HttpResponseMessage response1 = await client.GetAsync(url);
+                    Assert.Equal(HttpStatusCode.OK, response1.StatusCode);
+                    Assert.Equal(s_simpleContent, await response1.Content.ReadAsStringAsync());
+
+                    // Send second request.  Should reuse same connection.  
+                    // The server will close the connection, but HttpClient should retry the request.
+                    HttpResponseMessage response2 = await client.GetAsync(url);
+                    Assert.Equal(HttpStatusCode.OK, response1.StatusCode);
+                    Assert.Equal(s_simpleContent, await response1.Content.ReadAsStringAsync());
+                }
+            },
+            async server =>
+            {
+                await LoopbackServer.AcceptSocketAsync(server, async (s, stream, reader, writer) =>
+                {
+                    // Initial response
+                    await LoopbackServer.ReadWriteAcceptedAsync(s, reader, writer, s_simpleResponse);
+
+                    // Second response: Read request headers, then close connection
+                    await LoopbackServer.ReadWriteAcceptedAsync(s, reader, writer, "");
+                    s.Close();
```

Comment:
Perhaps it is already covered by other tests but it will be nice to have a test in which the server forces the client to re-use the same connection so we ensure pooling is working as expected.

Reply:
We do have that, though right now it only runs for ManagedHandler:
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/tests/FunctionalTests/ManagedHandlerTest.cs#L273